### DISCUSSION TITLE
How to use ChartJS with plugin chartjs-plugin-annotation?

### DISCUSSION THREAD

**Discussion by baxsmaxbax** [primefaces/community#187] [PrimeVue]
I couldn't find any example working in a component that uses ChatJS.

Most likely, the question is reduced to: "how to register a plugin".

**Comment by FlipWarthog** [comment]
See this example, which adds the chartjs-plugin-datalabels plugin: https://stackblitz.com/edit/mgfvrw?file=src%2FApp.vue

**Reply by baxsmaxbax** [reply]
Thank you very much for example. The key point turned out to be to import the plugin in the <script setup> block. I did the same thing, but without this nuance, nothing helped.
```
<template>
    <Chart type="line" :data="chart_data" :options="basicOptions" :height="130" :plugins="[ChartAnnotation]" />	
</template>
<script setup>
	import ChartAnnotation from 'chartjs-plugin-annotation';
</script>

<script>
import Chart from 'primevue/chart';
. . .
```
Now I see that the plugin has really connected and started working, because it generates errors :)

> Uncaught (in promise) TypeError: resolver is undefined

> Uncaught (in promise) Error: Canvas is already in use. Chart with ID '0' must be destroyed before the canvas with ID '' can be reused.

**Comment by FlipWarthog** [comment]
Hmm might an issue having two script tags. Try something like:
```vue
<template>
    <Chart type="line" :data="chart_data" :options="basicOptions" :height="130" :plugins="[ChartAnnotation]" />	
</template>

<script>
import Chart from 'primevue/chart';
import ChartAnnotation from 'chartjs-plugin-annotation';
. . .

**Reply by baxsmaxbax** [reply]
When I move import ChartAnnotation from 'chartjs-plugin-annotation'; to main script-block (without setup) - an error appears: "plugin is undefined".

**Reply by baxsmaxbax** [reply]
But now I have current error: "TypeError: resolver is undefined" from plugin

**Reply by FlipWarthog** [reply]
Are you able to reproduce the error in a Stackblitz? I could mess with it and see what I come up with

**Reply by movy** [reply]
This is still unresolved btw. Within 2 hours I could not figure out how it's done with PrimeVue so had to use VueChartJS to integrate a chart with a plugin (https://github.com/sgratzl/chartjs-chart-boxplot#esm-and-tree-shaking)

No matter what I tried with PrimeVue, I always get: 

```
chunk-Q7BMUYOJ.js?v=311491e8:7934 Uncaught (in promise) Error: "boxplot" is not a registered controller.
    at Registry._get (chunk-Q7BMUYOJ.js?v=311491e8:7934:13)
    at Registry.getController (chunk-Q7BMUYOJ.js?v=311491e8:7878:17)
    at Chart.buildOrUpdateControllers (chunk-Q7BMUYOJ.js?v=311491e8:8730:42)
    at Chart.update (chunk-Q7BMUYOJ.js?v=311491e8:8767:33)
    at new Chart (chunk-Q7BMUYOJ.js?v=311491e8:8537:12)
    at primevue_chart.js?v=311491e8:98:25
```

Repro:
https://stackblitz.com/edit/mgfvrw-3xiqfx?file=src%2FApp.vue

(the page is empty because it errors on setup, please see console).

I unearthed this commit: 
https://github.com/primefaces/primevue/blob/a6e06572d06b7ebc456eb28e382af726cf156347/components/lib/chart/Chart.vue#L34

Which supposedly exposes chart instance after the initialisation, but still was not able to register BoxPlot plugin.

Same idea works easy with VueChartJS:

```js
<template>
    <Bar
      v-if="chartData.datasets"
      :class="loading ? 'blurred' : ''"
      :data="chartData" 
      type="boxplot"
      :options="chartOptions"
    />  
</template>

<script setup>
import { Chart, LinearScale, CategoryScale } from 'chart.js'
import { BoxPlotController, BoxAndWiskers } from '@sgratzl/chartjs-chart-boxplot'
import { Bar } from 'vue-chartjs'

Chart.register(BoxPlotController, BoxAndWiskers, LinearScale, CategoryScale)
...
</script>
```

Still it's ugly, would be great to be able to register plugins from within PrimeVue.

**Reply by FlipWarthog** [reply]
@movy I think the issue is there are different kinds of plugins for Chart.js, e.g., `inline`/`registered`/`global`: https://www.chartjs.org/docs/latest/developers/plugins.html

See my example above, some plugins can definitely be used if they can be used inline, it may be that registering support is needed.

**Reply by movy** [reply]
So, to make it clear, currently there's no way to call `Chart.register()` from within PrimeVue chart component?

**Reply by FlipWarthog** [reply]
Not that I know of, unless you can hook in to the `loaded` emit and register that way, but that might be too late for when it's needed: https://primevue.org/chart/#api.chart.emits.loaded

Otherwise it would probably be a feature request to include additional info, like registers, for when the chart it loaded here: https://github.com/primefaces/primevue/blob/master/components/lib/chart/Chart.vue#L44-L60

**Reply by movy** [reply]
yes, `@loaded` seems too late to call anything, as I have tried this already and cannot see `.register()` method on `chart` returned by `loaded` event.
